xdsl/traits.py

```diff
@@ -476,28 +545,54 @@ class NoMemoryEffect(MemoryEffect):
     """
 
     @classmethod
-    def has_effects(cls, op: Operation) -> bool:
-        return False
+    def get_effects(cls, op: Operation) -> set[MemoryEffectKind]:
+        return set()
+
+
+class MemoryReadEffect(MemoryEffect):
+    """
+    A trait that signals that an operation always and only have read side effects.
```

Comment:
nit: we can drop the "and only" here.

```suggestion
    A trait that signals that an operation always have read side effects.
```

Reply:
"has" instead of "have", probably also